File: agent/commands-generator/src/main/java/org/matilda/commands/names/NameGenerator.java
package org.matilda.commands.names;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.TypeName;
import org.apache.commons.lang3.StringUtils;
import org.matilda.commands.info.CommandInfo;
import org.matilda.commands.info.ServiceInfo;

import javax.inject.Inject;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class NameGenerator {
    @Inject
    NameGenerator() {}

    public static String joinPackages(String... packages) {
        return joinPackages(Arrays.asList(packages));
    }

    public static String joinPackages(List<String> packages) {
        return packages.stream()
                .filter(part -> !part.isEmpty())
                .collect(Collectors.joining("."));
    }

    public static List<String> splitPackageName(String packageName) {
        return Arrays.asList(packageName.split("\\."));
    }

    public static final String MAIN_GENERATED_PACKAGE = "org.matilda.generated";
    public static final String COMMANDS_GENERATED_PACKAGE = joinPackages(MAIN_GENERATED_PACKAGE, "commands");
    public static final String RAW_COMMAND_CLASSES_PACKAGE = joinPackages(COMMANDS_GENERATED_PACKAGE, "raw");
    public static final String COMMANDS_MODULE_CLASS_NAME = "CommandsModule";
    public static final TypeName COMMANDS_MODULE_TYPE_NAME = ClassName.get(COMMANDS_GENERATED_PACKAGE,
            COMMANDS_MODULE_CLASS_NAME);
    public static final List<String> ORIGINAL_PACKAGE_PARTS = List.of("org", "matilda", "commands");

    public static class ServiceNameGenerator {
        private final ServiceInfo mServiceInfo;

        public ServiceNameGenerator(ServiceInfo serviceInfo) {
            mServiceInfo = serviceInfo;
        }

        public List<String> getFullNameParts() {
            return splitPackageName(mServiceInfo.fullName());
        }

        public String getServiceClassName() {
            List<String> nameParts = getFullNameParts();
            return nameParts.get(nameParts.size() - 1);
        }

        public List<String> getPackageParts() {
            List<String> nameParts = getFullNameParts();
            return nameParts.subList(0, nameParts.size() - 1);
        }

        private String getServiceRelativePackageName() {
            return joinPackages(getServiceRelativePackageNameParts());
        }

        private List<String> getServiceRelativePackageNameParts() {
            List<String> splitPackageName = getPackageParts();
            int equalParts = 0;
            while (equalParts < splitPackageName.size() && equalParts < ORIGINAL_PACKAGE_PARTS.size()) {
                if (!splitPackageName.get(equalParts).equals(ORIGINAL_PACKAGE_PARTS.get(equalParts))) {
                    break;
                }
                equalParts++;
            }
            return splitPackageName.subList(equalParts, splitPackageName.size());
        }

        public class CommandNameGenerator {
            private final CommandInfo mCommandInfo;

            public CommandNameGenerator(CommandInfo commandInfo) {
                mCommandInfo = commandInfo;
            }

            public String getRawCommandClassName() {
                return getServiceClassName() + StringUtils.capitalize(mCommandInfo.name()) + "Command";
            }

            public String getRawCommandPackageName() {
                return joinPackages(RAW_COMMAND_CLASSES_PACKAGE, getServiceRelativePackageName());
            }

            public TypeName getRawCommandTypeName() {
                return ClassName.get(getRawCommandPackageName(), getRawCommandClassName());
            }

            public String getFullCommandName() {
                return getServiceRelativePackageNameParts().stream()
                        .map(StringUtils::capitalize)
                        .collect(Collectors.joining()) + getRawCommandClassName();
            }
        }
    }

    public ServiceNameGenerator.CommandNameGenerator forCommand(CommandInfo commandInfo) {
        return new ServiceNameGenerator(commandInfo.service()).new CommandNameGenerator(commandInfo);
    }
}

File: agent/commands-generator/src/main/java/org/matilda/commands/names/NameGenerator.kt
package org.matilda.commands.names

import com.squareup.javapoet.ClassName
import org.apache.commons.lang3.StringUtils
import org.matilda.commands.info.CommandInfo
import org.matilda.commands.info.ServiceInfo
import org.matilda.commands.java.JavaProperties
import org.matilda.commands.python.PythonClassName
import org.matilda.commands.python.PythonProperties
import org.matilda.commands.utils.Package
import org.matilda.commands.utils.Package.Companion.fromString
import org.matilda.commands.utils.Package.Companion.joinPackages
import org.matilda.commands.utils.toSnakeCase
import javax.inject.Inject

class NameGenerator @Inject internal constructor() {
    @Inject
    lateinit var mPythonProperties: PythonProperties

    @Inject
    lateinit var mJavaProperties: JavaProperties

    private val javaMainPackage
        get() = mJavaProperties.javaMainPackage

    private val javaMainGeneratedPackage
        get() = javaMainPackage.subpackage("generated")

    val commandsGeneratedPackage
        get() = javaMainGeneratedPackage.subpackage("commands")

    val rawCommandClassesPackage
        get() = commandsGeneratedPackage.subpackage("raw")
    val dependenciesClassesPackage
        get() = commandsGeneratedPackage.subpackage("dependencies")
    val dynamicServiceConvertersClassesPackage
        get() = commandsGeneratedPackage.subpackage("converters")
    val commandRegistryFactoriesPackage
        get() = commandsGeneratedPackage.subpackage("registryFactories")
    val javaServiceProxiesPackage
        get() = commandsGeneratedPackage.subpackage("proxies")

    val commandsModuleClassName: ClassName
        get() = ClassName.get(commandsGeneratedPackage.packageName, "CommandRegistryModule")

    val servicesModuleClassName: ClassName
        get() = ClassName.get(commandsGeneratedPackage.packageName, "ServicesModule")

    val pythonGeneratedCommandsPackage: Package
        get() = mPythonProperties.pythonGeneratedPackage.subpackage("commands")

    val pythonGeneratedServicesContainerPackage: Package
        get() = mPythonProperties.pythonGeneratedPackage.subpackage("services")

    inner class ServiceNameGenerator(private val mServiceFullName: String) {
        private val fullNamePackage: Package
            get() = fromString(mServiceFullName)
        val serviceClassName: String
            get() = fullNamePackage.lastPart

        val serviceSnakeCaseName: String
            get() = serviceClassName.toSnakeCase()

        val serviceFullClassName: PythonClassName
            get() = PythonClassName.createFromParentPackageAndClass(pythonGeneratedCommandsPackage, serviceClassName)

        private val servicePackage: Package
            get() = fullNamePackage.withoutLastPart()
        private val serviceRelativePackage: Package
            get() = servicePackage.removeCommonPrefixFrom(javaMainPackage)
        val commandRegistryFactoryClassName: ClassName
            get() = ClassName.get(
                joinPackages(commandRegistryFactoriesPackage, serviceRelativePackage).packageName,
                serviceClassName + "CommandRegistryFactory"
            )
        val commandRegistryFactoryPythonClassName: PythonClassName
            get() = PythonClassName.createFromParentPackageAndClass(
                pythonGeneratedCommandsPackage.subpackage("registry_factories"),
                "${serviceClassName}CommandRegistryFactory"
            )
        val javaServiceProxyClassName: ClassName
            get() = ClassName.get(
                joinPackages(javaServiceProxiesPackage, serviceRelativePackage).packageName,
                serviceClassName + "Proxy"
            )
        val serviceProxyClassName: PythonClassName
            get() = PythonClassName.createFromParentPackageAndClass(
                pythonGeneratedCommandsPackage.subpackage("proxies"),
                "${serviceClassName}Proxy"
            )
        val javaServiceProxyFactoryClassName: ClassName
            get() = ClassName.get(
                joinPackages(javaServiceProxiesPackage, serviceRelativePackage).packageName,
                serviceClassName + "ProxyFactory"
            )
        val pythonServiceProxyFactoryClassName: PythonClassName
            get() = PythonClassName.createFromParentPackageAndClass(
                pythonGeneratedCommandsPackage.subpackage("proxies"),
                serviceClassName + "ProxyFactory"
            )
        val dependenciesClassName: ClassName
            get() = ClassName.get(
                joinPackages(dependenciesClassesPackage, serviceRelativePackage).packageName,
                serviceClassName + "Dependencies"
            )

        val dependenciesPythonClassName: PythonClassName
            get() = PythonClassName.createFromParentPackageAndClass(
                pythonGeneratedCommandsPackage.subpackage("dependencies"),
                "${serviceClassName}Dependencies"
            )

        val dynamicServiceConverterClassName: ClassName
            get() = ClassName.get(
                joinPackages(dynamicServiceConvertersClassesPackage, serviceRelativePackage).packageName,
                serviceClassName + "Converter"
            )

        val dynamicServiceConverterPythonClassName: PythonClassName
            get() = PythonClassName.createFromParentPackageAndClass(
                pythonGeneratedCommandsPackage.subpackage("converters"),
                serviceClassName + "Converter"
            )

        inner class CommandNameGenerator(private val mCommandInfo: CommandInfo) {
            val rawCommandClassName: ClassName
                get() = ClassName.get(
                    joinPackages(rawCommandClassesPackage, serviceRelativePackage).packageName,
                    serviceClassName + StringUtils.capitalize(mCommandInfo.name) + "Command"
                )
            val rawCommandPythonClassName: PythonClassName
                get() = PythonClassName.createFromParentPackageAndClass(
                    pythonGeneratedCommandsPackage.subpackage("raw"),
                    rawCommandClassName.simpleName()
                )
            val fullCommandName: String
                get() = serviceRelativePackage.parts.joinToString(separator = "") {
                    str -> StringUtils.capitalize(str)
                } + rawCommandClassName.simpleName()

            val snakeCaseName: String
                get() = mCommandInfo.name.toSnakeCase()
        }
    }

    fun forService(serviceInfo: ServiceInfo) = forService(serviceInfo.fullName)
    fun forService(fullName: String) = ServiceNameGenerator(fullName)

    fun forCommand(commandInfo: CommandInfo) = forService(commandInfo.service).CommandNameGenerator(commandInfo)

    companion object {
        const val SERVICES_CONTAINER_CLASS_NAME = "Services"
    }
}
